man/title.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH TITLE 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
title \- generates information about a specific movie\-title.
.SH SYNOPSIS
.B
title \-t "title" <output\-options> [\-t "title" <output\-options>] <misc options>
.SH DESCRIPTION
The
.B title
program searches the database for all the information relating
to a specific title. The basic option is
.B
-t "title"
which specifies the movie to search for.
.SH OPTIONS
Options are divided into two groups:
.B output-options
and
.BR "miscellaneous options" .
.SS OUTPUT-OPTIONS
.TP
.B \-aka
Add any alternative titles.  ("aka" means Also Known As)
.TP
.B \-mrr
Add movie ratings report data.
.TP
.B \-yr
Add year of release and any title attributes.
.TP
.B \-pl
Add plot summary.
.TP
.B \-chr
Add character names to cast lists.
.TP
.B \-trivia
Add movie trivia, crazy credits, goofs, quotes, soundtrack information,
certificates, color information, countries, genres, production companies,
running times and sound mix.
.TP
.B \-full
Combines all of the above options.
.SS EXAMPLES ON OUTPUT-OPTIONS
.TP
List all information for "2001: A Space Odyssey"
.B
title \-t "2001: A Space Odyssey" \-full
.TP
List details of both the 1962 and 1991 version of "Cape Fear"
.B
title \-t "Cape Fear"
.TP
List details for only the 1991 version of "Cape Fear"
.B
title \-t "Cape Fear (1991)"
.SS MISCELLANEOUS OPTIONS
.TP
.B \-i
Case sensitive search: normally searches are not case sensitive.
.TP
.B \-s
Substring: match on substring rather than whole title.
.SS EXAMPLES ON MISCELLANEOUS OPTIONS
title -t "Thin Man" -full -s
.SH Multiple Title Searches
Any number of
.B \-t
options may be combined in the same query to generate a 
list of people that worked on each and every one of the movies.
.SS EXAMPLES ON Multiple Title Searches
title -t "Ghostbusters" -t "Ghostbusters II"


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/title.c
/*============================================================================
 *
 *  Program: title.c 
 *
 *  Version: 3.7
 *
 *  Purpose: search databases for titles
 *
 *  Author:  C J Needham <col@imdb.com>     
 *
 *  Copyright (c) 1996-1999 The Internet Movie Database Inc.
 *
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 * 
 *  Options:
 *    -t <title>  search for details on title
 *    -s          treat title as substring and return all matches
 *    -i          case-sensitive search for -s
 *   
 *    -yr         display year of release
 *    -mrr        display movie ratings report info
 *    -aka        search alternative title database
 *    -pl         add plot description
 *    -lit        add literature 
 *    -bus        add business 
 *    -ld         add laserdisc
 *    -trivia     add any trivia, crazy credits, quotes and goofs
 *
 *    -full       short hand for -yr -mrr -pl -trivia -lit -bus -ld
 *
 *============================================================================
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "log.h"
#include "moviedb.h"
#include "dbutils.h"
#include "titlesearch.h"
#include "display.h"

#define TITLE_USAGE1 "usage: title [-t <title>]+ [-i -s]"
#define TITLE_USAGE2 "             [-aka -mrr -yr -pl -trivia -lit -bus -ld] [-full]"

int main ( int argc, char **argv )
{
  int   err = FALSE ;
  int   i, j ;
  int   fullopt = FALSE ;
  int   tidy = FALSE ;
  int   trivopts [ NO_OF_TRIV_LISTS ] ;
  int   titleInfoOpts [ NO_OF_TITLE_INFO_LISTS ] ;
  struct titleSearchRec *combined = NULL, *tchain = NULL ;
  struct titleSearchOptRec options = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 } ;

  logProgram ( argv [ 0 ] ) ;
  options . yearMatch = TRUE ;

  if ( argc == 1 )
    err = TRUE ;

  for ( i = 0 ; i < NO_OF_TRIV_LISTS ; i++ )
    trivopts [ i ] = FALSE ;

  for ( i = 0 ; i < NO_OF_TITLE_INFO_LISTS ; i++ )
    titleInfoOpts [ i ] = FALSE ;

  for ( i = 1; i < argc; i++ )
    if ( strcmp ( "-t", argv[i] ) == 0 )
      if ( ++i < argc )
        tchain = addTitleSearchRec ( tchain, argv [ i ] ) ;
      else
        err = TRUE ;
    else if ( strcmp ( "-i", argv [ i ] ) == 0 )
      options . casesen = TRUE ;
    else if ( strcmp ( "-pl", argv [ i ] ) == 0 )
    {
      options . plotopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-bus", argv [ i ] ) == 0 )
    {
      options . businessopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-ld", argv [ i ] ) == 0 )
    {
      options . ldopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-nold", argv [ i ] ) == 0 )
    {
      options . ldopt = FALSE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-lit", argv [ i ] ) == 0 )
    {
      options . litopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-ym", argv [ i ] ) == 0 )
    {
      options . yearMatch = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-exact", argv [ i ] ) == 0 )
      options . yearMatch = FALSE ;
    else if ( strcmp ( "-s", argv [ i ] ) == 0 )
    {
      options . substring = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-mrr", argv [ i ] ) == 0 )
    {
      options . mrropt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-yr", argv [ i ] ) == 0 )
    {
      options . yropt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-trivia", argv [ i ] ) == 0 )
    {
      for ( j = 0 ; j < NO_OF_TRIV_LISTS ; j++ )
        trivopts [ j ] = TRUE ;
      for ( j = 0 ; j < NO_OF_TITLE_INFO_LISTS ; j++ )
        titleInfoOpts [ j ] = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-aka", argv [ i ] ) == 0 )
    {
      options . akaopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-link", argv [ i ] ) == 0 )
    {
      options . linkopt = TRUE ;
      tidy = TRUE ;
    }
    else if ( strcmp ( "-full", argv [ i ] ) == 0 )
    {
      for ( j = 0 ; j < NO_OF_TRIV_LISTS ; j++ )
        trivopts [ j ] = TRUE ;
      for ( j = 0 ; j < NO_OF_TITLE_INFO_LISTS ; j++ )
        titleInfoOpts [ j ] = TRUE ;
      options . mrropt = TRUE ;
      options . yropt = TRUE ;
      options . plotopt = TRUE ;
      options . businessopt = TRUE ;
      options . ldopt = TRUE ;
      options . litopt = TRUE ;
      options . akaopt = TRUE ;
      options . linkopt = TRUE ;
      tidy = TRUE ;
      fullopt = TRUE ;
    }
    else
    {
      (void) fprintf ( stderr, "title: unrecognised option %s\n", argv[i] ) ;
      err = TRUE ;
    }

  if ( err || tchain == NULL )
    moviedbUsage ( TITLE_USAGE1, TITLE_USAGE2, NULL, NULL, NULL, NULL ) ;

  if ( tchain -> next != NULL )
  {
    if ( options . substring )
      moviedbError ("title: -s not available with more than one search" ) ;
    else if ( fullopt )
      moviedbError ( "title: -full not available with more than one search" ) ;
    else if ( options . mrropt )
      moviedbError ( "title: -mrr not available with more than one search" ) ;
    else if ( options . yropt )
      moviedbError ( "title: -yr not available with more than one search" ) ;
    else if ( options . plotopt )
      moviedbError ( "title: -pl not available with more than one search" ) ;
    else if ( options . businessopt )
      moviedbError ( "title: -bus not available with more than one search" ) ;
    else if ( options . ldopt )
      moviedbError ( "title: -ld not available with more than one search" ) ;
    else if ( options . litopt )
      moviedbError ( "title: -lit not available with more than one search" ) ;
    else if ( options . akaopt )
      moviedbError( "title: -aka not available with more than one search" ) ;
    else if ( trivopts [ 0 ] )
      moviedbError( "title: -trivia not available with more than one search" ) ;
    else if ( titleInfoOpts [ 0 ] )
      moviedbError( "title: -info not available with more than one search" ) ;
  }

  if ( tchain != NULL && tchain -> next != NULL )
  {
    addTitleChainOpts ( tchain, options ) ;
    addTitleChainTrivOpts ( tchain, trivopts ) ;
    addTitleChainTitleInfoOpts ( tchain, titleInfoOpts ) ;
    processTitleSearch ( tchain ) ;
    combined = combineTitleResults ( tchain ) ; 
    displayTitleSearchResults ( combined, FALSE ) ; 
    freeTitleChain ( combined ) ; 
  }
  else 
  { 
    addTitleChainOpts ( tchain, options ) ;
    addTitleChainTrivOpts ( tchain, trivopts ) ;
    addTitleChainTitleInfoOpts ( tchain, titleInfoOpts ) ;
    processTitleSearch ( tchain ) ;
    displayTitleMatches ( tchain ) ;
    displayTitleSearchResults ( tchain, tidy ) ; 
  } 

  freeTitleChain ( tchain ) ;

  return ( 0 ) ;
}
